src/seq/demultiplexer.jl

```diff
@@ -0,0 +1,336 @@
+    println(io, "  number of barcodes:", length(demultiplexer.barcodes))
+      print(io, "  number of correctable errors: ", length(demultiplexer.tries) - 1)
+end
+
+"""
+    Demultiplexer(barcodes::Vector{DNASequence};
+                  n_max_errors::Integer=1,
+                  distance::Symbol=:hamming)
+
+Create a demultiplexer object from `barcodes`.
+
+# Arguments
+* `barcodes`: the sorted list of DNA barcodes; integer IDs are assigned in this order.
+* `n_max_errors=1`: the number of maximum correctable errors.
+* `distance=:hamming`: the distance metric (`:hamming` or `:levenshtein`).
+"""
+function Demultiplexer(barcodes::Vector{DNASequence};
+                       n_max_errors::Integer=1,
+                       distance::Symbol=:hamming)
+    if !issorted(barcodes)
+        error("barcodes must be sorted")
+    elseif n_max_errors < 0
+        error("n_max_errors must be non-negative")
+    elseif distance ∉ (:hamming, :levenshtein)
+        error("distance must be either :hamming or :levenshtein")
+    end
+
+    # check barcodes
+    for barcode in barcodes
+        for i in 1:endof(barcode)
+            if isambiguous(barcode[i])
+                error("barcode must be umambiguous")
+            end
+        end
+    end
+
+    # build trie trees for each error level (0-n_max_errors)
+    tries = BarcodeTrie[]
+    for m in 0:n_max_errors
+        # generate "erroneous" barcodes
+        barcodes′ = DNASequence[]
+        ids = Int[]
+        for (i, barcode) in enumerate(barcodes)
+            if distance == :hamming
+                circle = hamming_circle(barcode, m)
+            elseif distance == :levenshtein
+                circle = levenshtein_circle(barcode, m)
+            else
+                # unreachable: already checked above
+                assert(false)
+            end
+            append!(barcodes′, circle)
+            append!(ids, collect(repeated(i, length(circle))))
+        end
+        ord = sortperm(barcodes′)
+        push!(tries, BarcodeTrie(barcodes′[ord], ids[ord]))
+    end
+
+    return Demultiplexer(tries, barcodes, distance)
+end
+
+"""
+    demultiplex(demultiplexer::Demultiplexer, seq::Sequence, linear_search_fallback::Bool=false)
+
+Return a barcode index that matches `seq` with least errors.
+
+The upper limit of the number of maximum errors is bounded by the `n_max_errors`
+parameter of `demultiplexer`. When `linear_search_fallback` is `true`, this
+function tries to find the best matching barcode using linear search.
+"""
+function demultiplex(demultiplexer::Demultiplexer, seq::Sequence, linear_search_fallback::Bool=false)
+    if eltype(seq) != DNANucleotide
+        error("sequence must be a DNA sequence")
+    end
+    for trie in demultiplexer.tries
+        i = findbarcode(trie, seq)
+        if i != 0
+            return i
+        end
+    end
+    if !linear_search_fallback
+        # not found
+        return 0
+    end
+
+    # find the best matching barcode using linear search
```

Comment:
Ah, it probably is to a point, but clearly if you have `NNNN` in the read, this becomes random assignment (well arbitrary assignment at least). Provided that we return the number of mismatches, then I say it's fine. It may be worth capping the maximum distance at say 1/2 the barcode length or something, beyond which the linear search assignment won't reflect the most likely original barcode as they're too different.